### ISSUE TITLE
[question] Emscripten having a fixed NodeJS dependency is problematic

### ISSUE THREAD

**Issue by johannes-wolf** [conan-io/conan-center-index#23088]
See: https://github.com/conan-io/conan-center-index/blob/d4d9670d91ba0abfb232d9233170b229fa6a1289/recipes/emsdk/all/conanfile.py#L32

A tool having a dependency on a fixed version of Node.js is problematic. If I want to use `emsdk` via Conan, I am forced to use Node.js 16.3.0, which I do not want to. What is the solution for bringing my own Node.js version? Would adding an option to specify the Node.js package version be ok? Something like: `self.requires("nodejs/" + self.options.nodejs_version)`.

PR: #23096

**Comment by uilianries** [comment]
@johannes-wolf You should use [replace_requires](https://docs.conan.io/2/reference/config_files/profiles.html#replace-requires) (NOTE: Still experiment feature) to override the version for nodejs.

You would need:

- In profile add:

```
[replace_requires]
nodejs/*: nodejs/16.20.2
```

Or what version do you want to be installed. It will generate the follow output (Conan 2.1):

```
nodejs/16.20.2: Not found in local cache, looking in remotes...
nodejs/16.20.2: Checking remote: conancenter
nodejs/16.20.2: Downloaded recipe revision b7dcca884fe50f9d0af67128c44359a8
Graph root
    conanfile.txt: /private/tmp/foo/conanfile.txt
Requirements
    emsdk/3.1.50#4978d062889977c9631481e286be8589 - Cache
    nodejs/16.20.2#b7dcca884fe50f9d0af67128c44359a8 - Downloaded (conancenter)
Replaced requires
    nodejs/16.3.0: nodejs/16.20.2
```

As you can see, emsdk uses nodejs/16.3.0, but with replace_requires we replaced by 16.20.2. It will return an error about missing package, as there is not emsdk with such combination, but no conflict error will be present.

**Comment by johannes-wolf** [comment]
Sadly, this does not work.

I have the following profile:
```
[settings]
os=Emscripten
arch=wasm
compiler=clang
compiler.version=15
compiler.libcxx=libc++
compiler.cppstd=20
build_type=Release

[tool_requires]
emsdk/3.1.50

[replace_requires]
nodejs/*: nodejs/16.20.2

[replace_tool_requires]
nodejs/*: nodejs/16.20.2
```

I even placed the override in `*_requires` and `*_tool_requires`: nothing gets overridden.

The `self.requires(...)` option you've mentioned does not work because `emsdk` is a `tool_requires`, which can not be overridden.

**Comment by uilianries** [comment]
@johannes-wolf Which Conan version are you running? I tested with 2.1 (latest) and worked.

**Comment by johannes-wolf** [comment]
> @johannes-wolf What Conan version are you running? I tested with 2.1 (latest) and worked.

With 2.1 installed via pip.

**Comment by johannes-wolf** [comment]
I've added a simple example:
[conan-replace-test.tar.gz](https://github.com/conan-io/conan-center-index/files/14607667/conan-replace-test.tar.gz)

If I install using:
`conan install . -of build -pr:h profile.profile -pr:b default`

The log shows `nodejs/16.3.0`, although the profile specifies the override.

**Comment by uilianries** [comment]
@johannes-wolf Please, take a look in the issue https://github.com/conan-io/conan/issues/15749. There is a reference to Conan blog explaining that feature in detail.

**Comment by johannes-wolf** [comment]
TL;DR: You have to add the override (using `replace_requires`!) to your build profile (`-pr:b`).
Example:
```toml
include(default)

[replace_requires]
nodejs/*: nodejs/16.20.2
```

**Comment by rturrado** [comment]
So are we going to do anything with the `emsdk` Conan recipe? Can't we not just update the recipe to point to `nodejs/16.20.2`? (not sure if that's the solution to `emsdk/3.1.51`+ not building).

Because that latest comment from @johannes-wolf works only as a workaround for `emsdk/3.1.50` Conan package users that are getting errors due to `emsdk` requesting `nodejs/16.20.3`, right?
